src/forcing/ForcingsEngine.cpp

```diff
@@ -0,0 +1,239 @@
+#include "ForcingsEngine.hpp"
+
+namespace data_access {
+
+std::unordered_map<std::string, std::unique_ptr<ForcingsEngine>> ForcingsEngine::instances_{};
+
+ForcingsEngine::ForcingsEngine(const std::string& init, size_type time_start, size_type time_end)
+  : time_start_(std::chrono::seconds{time_start})
+  , time_end_(std::chrono::seconds{time_end})
+{
+    check_runtime_requirements();
+
+    bmi_ = std::make_unique<models::bmi::Bmi_Py_Adapter>(
+        "ForcingsEngine",
+        init,
+        "NextGen_Forcings_Engine.BMIForcingsEngine",
+        true,
+        true,
+        utils::getStdOut()
+    );
+    
+    time_step_ = std::chrono::seconds{static_cast<int64_t>(bmi_->GetTimeStep())};
+    var_outputs_ = bmi_->GetOutputVarNames();
+    var_outputs_.erase(var_outputs_.begin()); // Erase "CAT-ID" from outputs
+    const auto time_dim = static_cast<size_type>((time_end_ - time_start_) / time_step_);
+    const auto id_dim   = static_cast<size_type>(bmi_->GetVarNbytes("CAT-ID") / bmi_->GetVarItemsize("CAT-ID"));
+    const auto var_dim  = var_outputs_.size();
+
+    bmi_->Update();
+    time_current_index_++;
+    
+    const auto* ptr = static_cast<int*>(bmi_->GetValuePtr("CAT-ID"));
+    var_divides_ = std::vector<int>(ptr, ptr + id_dim);
+    var_cache_ = decltype(var_cache_){{ time_dim, id_dim, var_dim }};
+
+    // Cache initial iteration
+    const auto size = var_divides_.size();
+    for (size_type vi = 0; vi < var_outputs_.size(); vi++) {
+        // Get current values for each variable
+        const auto var_span = boost::span<const double>{
+            static_cast<double*>(bmi_->GetValuePtr(var_outputs_[vi])),
+            size
+        };
+
+        for (size_type di = 0; di < size; di++) {
+            // Set current variable value for each divide
+            var_cache_.at({{time_current_index_, di, vi}}) = var_span[di];
+        }
```

Comment:
This may be a pre-mature optimization, but I think it would be good to have some sort of ranged `set` operation in the `mdarray` itself that can do this a bit more efficiently. If nothing else, the "store values from a new time step" functionality is repeated below inside `at`, so it should at least be factored out to its own helper function

Reply:
I factored out the helper function in [6edd99a](https://github.com/NOAA-OWP/ngen/pull/720/commits/6edd99acbfb710ba2294225d3e65d893c04d90ff) for now. I'm going to hold off on trying to implement a ranged set op in mdarray and probably do that in a separate PR